Create Entities for Courses Feature

ajaygandecha opened the pull request:
This major makes progress on adding all of the necessary entities for the courses feature to function. This includes including courses, terms (semesters), course sections, and the necessary association tables to connect users to sections, both as students and as instructors.

## Implementation

The entities follow this basic diagram:

![Image from iOS](https://github.com/unc-csxl/csxl.unc.edu/assets/17516747/d823cfbf-6afb-4cea-8634-2048acb4415c)

The addition of relationships between the `section` and `room` table can wait until it becomes necessary to add it.

Note that to keep with consistency of the *coworking* feature, all table names in the database have been added with a `courses__` prefix.

ajaygandecha commented:
@KrisJordan / @jadekeegan  / @atoneyd  / @briannata main pieces here are if the entity-level relationships and table names look good!

ajaygandecha commented:
@KrisJordan, ready for review. The second join table has been removed!

KrisJordan commented on backend/entities/courses/user_section_entity.py:
```diff
@@ -0,0 +1,42 @@
+    # 1 = Undergraduate Teaching Assistant
+    # 2 = Graduate Teaching Assistant
+    # 3 = Main Instructor
+    member_type: Mapped[int] = mapped_column(Integer)
```
Let's use an enum rather than just an integer here:
<https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Enum>

Perhaps:

STUDENT
UTA
GTA
INSTRUCTOR

KrisJordan reviewed (changes requested):
Given the long-lasting importance of getting these entities modeled as ideally as possible, I'm digging in on some details here. In general, though, I think this is very close and in the right direction!

ajaygandecha commented on backend/entities/__init__.py:
```diff
@@ -14,6 +14,7 @@
 
 
 from .entity_base import EntityBase
+from .courses import *
```
@KrisJordan thank you for pointing this out! This was actually a change I had to make because for some reason, the entities were not being recognized in the relationship field with `UserEntity`. If we were to not include this relationship at the entity level though (as might be suggested below), then we can go ahead and remove this.

ajaygandecha replied:
Removed this in the latest commit.

ajaygandecha commented on backend/entities/courses/course_entity.py:
```diff
@@ -0,0 +1,34 @@
+    # Course properties (columns in the database table)
+
+    # Unique ID for the course
+    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
```
I agree! Implementing the second approach.

ajaygandecha commented on backend/entities/courses/section_entity.py:
```diff
@@ -0,0 +1,39 @@
+    # Unique ID for the section
+    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
+    # Code of the section (for example, COMP 100-003's code would be "003")
+    code: Mapped[str] = mapped_column(String, default="")
```
Agree, should be `number`!

ajaygandecha commented on backend/entities/user_entity.py:
```diff
@@ -53,6 +53,13 @@ class UserEntity(EntityBase):
     permissions: Mapped["PermissionEntity"] = relationship(back_populates="user")
 
+    # NOTE: This field establishes a many-to-many relationship between the user and course section tables.
+    course_sections: Mapped[list["SectionEntity"]] = relationship(
+        secondary="courses__user_section",
+        back_populates="students",
+        cascade="all, delete",
+    )
```
@KrisJordan the only thing with this is that I need to remove the back_populated `members` field in the `SectionEntity`. I think that this is fine. As you mentioned, the event entity and user entity do not contain these relationship fields.

ajaygandecha commented on backend/entities/courses/course_entity.py:
```diff
@@ -0,0 +1,34 @@
+
+    # NOTE: This field establishes a one-to-many relationship between the course and section tables.
+    sections: Mapped[list["SectionEntity"]] = relationship(
+        back_populates="course", cascade="all,delete"
```
Agree with this too - same issue I mentioned above arises with `backend.script.delete_database` / `reset_demo`, though.

KrisJordan replied:
Let's keep it for convenience while developing now, but create a card for figuring out how to improve the reset demo script to avoid this. I can look into it. In MySQL there was a way to disable FK constraints in a transaction, until the transaction went to be completed. Maybe Postgres has a similar feature.

ajaygandecha commented:
Thank you for the thorough review here @KrisJordan! I implemented a lot of the changes that were suggested. Just a few key questions to summarize from what I wrote in the comments:
- I deleted the relationship fields in `UserEntity` and `SectionEntity` respectively - confirming this is good?
- Do we want to change the name of the course ID field? Is the primary key of terms being `shorthand` okay?
- How should we handle the `delete_database` / `reset_demo` script if we do not include cascading?

ajaygandecha commented on backend/entities/courses/section_entity.py:
```diff
@@ -0,0 +1,39 @@
+    # Section properties (columns in the database table)
+
+    # Unique ID for the section
+    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
```
@KrisJordan if we wanted to be extra.. UNC does assign IDs to each course section in ConnectCarolina specifically! We could use these 👀

ajaygandecha replied:
Also makes it easy if we wanted to make section data population scriptable. Think this is possible. May also just be perfectly fine to have a admin-level UI for making entering section data easier.

ajaygandecha commented on backend/entities/courses/user_section_entity.py:
```diff
@@ -0,0 +1,49 @@
+__license__ = "MIT"
+
+
+class CourseSectionMembershipType(Enum):
```
We can rename this if we want, since it is kind of long. Since the enum may be used outside of the entity though, I kept `CouseSection...` instead of just `Section...`

KrisJordan replied:
How about `CourseRole` or `RosterRole` as alternates?

ajaygandecha replied:
Like it! I might favor `RosterRole` because `CourseRole` might imply a relationship to courses rather than sections.

KrisJordan commented on backend/entities/courses/term_entity.py:
```diff
@@ -0,0 +1,33 @@
+
+    # NOTE: This field establishes a one-to-many relationship between the term and section tables.
+    course_sections: Mapped[list["SectionEntity"]] = relationship(
+        back_populates="term", cascade="all,delete"
```
👍👍 same as above comment. Keep for now while in dev.

KrisJordan commented:
Good point, this clarifies two things:

1. There's no legit concern on eagerly loading (without asking for it) to-many relations, this was my tired brain being silly.
2. The actual design decision is more about "needs to know" and expected uses. In our architecture, we use entities primarily to capture the database representations and have some niceties for persisting data in and out. Our service layer captures the more interesting business logic. It's a no-brainer to model a SQLAlchemy `relationship` on the entities who own the foreign key, the one side, but the question of whether it's worth having on the _many_ side seems to come down to how nuanced the use cases are for loading the many. Your example is a great case: how often do we want to load all 300 students (without pagination, concern for instructor vs. gta vs. student, etc)? I imagine that's exceedingly rare and the logic for how to expose the students in a section is better pushed up to the service layer. So I retract my recommendation and concur with yours. Not having the `relationship` in the entity is a good indication that the service layer will have more interesting logic/concerns for loading the data in the relationship rather than just "load all."

This lgtm to proceed!

KrisJordan commented on backend/entities/courses/course_entity.py:
```diff
@@ -0,0 +1,36 @@
+    # Title or name for the course
+    title: Mapped[str] = mapped_column(String, default="")
+    # Course description for the course
+    course_description: Mapped[str] = mapped_column(String, default="")
```
Yes, good call.

ajaygandecha commented on backend/entities/courses/course_entity.py:
```diff
@@ -0,0 +1,36 @@
+    # Examples: COMP110, COMP283H
+    id: Mapped[str] = mapped_column(String(9), primary_key=True)
+    # Subject for the course (for example, the subject of COMP 110 would be COMP)
+    subject_code: Mapped[str] = mapped_column(String, default="")
```
@KrisJordan do we want to provide a string length max for the code and number?

KrisJordan reviewed (approved):
LGTM with the minor change requested of adding max lengths to the fields you suggested!
